src/MSBuild/TerminalLogger/TerminalLogger.cs

```diff
@@ -635,28 +635,38 @@ private bool IsImmediateMessage(string message)
     /// </summary>
     private void ErrorRaised(object sender, BuildErrorEventArgs e)
     {
-        var buildEventContext = e.BuildEventContext;
-        if (buildEventContext is not null && _projects.TryGetValue(new ProjectContext(buildEventContext), out Project? project))
-        {
-            string message = EventArgsFormatting.FormatEventMessage(
+        BuildEventContext? buildEventContext = e.BuildEventContext;
+        Project? project = null;
+        bool isTrackedProject = buildEventContext is not null && _projects.TryGetValue(new ProjectContext(buildEventContext), out project);
+        string message = EventArgsFormatting.FormatEventMessage(
                 category: AnsiCodes.Colorize("error", TerminalColor.Red),
                 subcategory: e.Subcategory,
                 message: e.Message,
                 code: AnsiCodes.Colorize(e.Code, TerminalColor.Red),
                 file: HighlightFileName(e.File),
-                projectFile: null,
+
+                // for the tracked projects the project file name is included in the final output result.
+                projectFile: isTrackedProject ? null : e.ProjectFile ?? string.Empty,
```

Comment:
Shouldn't it always have the project, if the error message itself does? Doesn't seem restricted to the tracked project entry points.